### PR TITLE
[Traceable FSDP2] Add `aot_eager` backend E2E tests for transformer model

### PR DESCRIPTION
**Pull request by yf225**
This PR adds Traceable FSDP2 `aot_eager` backend E2E tests for simple MLP as well as transformer model.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129157
* #129203

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

### CODE REVIEW

File: torch/testing/_internal/llama_model.py [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+from dataclasses import dataclass
+from typing import Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+
+@dataclass
+class ModelArgs:
```
Comment (awgu) [inline torch/testing/_internal/llama_model.py:13]: For landing: I think we can reuse the `Transformer` and `ModelArgs` from `common_dtensor.py`.
https://github.com/pytorch/pytorch/blob/f2f4dde2d37d2e2e6c76cdeeeea376cc7ffb8016/torch/testing/_internal/distributed/_tensor/common_dtensor.py#L172

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 2]
```diff
@@ -63,7 +77,28 @@ def patched_trace_rules_check(*args, **kwargs):
             self.assertTrue(trace_rules_check_count > 0)
 
 
+def cleanup_fsdp(model):
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_compile.py:80]: Is it insufficient to just del the module(s)? I would hope the owned tensors get garbage collected then without having to manually `resize_(0)`.

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 3]
```diff
@@ -100,6 +135,184 @@ def f(x):
...
+
+            model_compiled = torch.compile(model, backend=backend, fullgraph=True)
+            res = run_all_iters(
+                model_compiled, optim, compiled_autograd_backend=backend
+            )
+            cleanup_fsdp(model)
+            del model_compiled
+            del model
+            optim.zero_grad(set_to_none=True)
+            del optim
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_compile.py:192]: For these, I wonder why we need to `del` explicitly? I would expect everything to get GC'd after this function returns.

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 2 above)
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:80]: A few weeks ago when I debug the memory usage issue, it isn't quite the case - `gc.collect` seems not able to free the owned tensors after the module is `del`-ed.

Alternatively we can remove this part and I can visit it again when I get to landing the memory tests for Inductor.

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 3 above)
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:192]: yes we probably don't need to do the `del`s, although the `resize_(0)` might be more fundamental (I can remove all these and investigate again when I get to the memory tests later)

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 2 above)
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py]: Removed this function and will visit again in the future

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 3 above)
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py]: Removed all `del`s

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 4]
```diff
@@ -100,6 +113,172 @@ def f(x):
...
+        hidden_dim = 16
+
+        def model_init_fn():
+            torch.manual_seed(0)
+            fsdp_config = {}
+            mesh = init_device_mesh("cuda", (self.world_size,))
+            model_args = ModelArgs(
+                dim=hidden_dim,
+                n_layers=2,
+                n_heads=1,
```
[the remark this answers is not in the document]
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:237]: changed to default args

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 5]
```diff
@@ -100,6 +113,172 @@ def f(x):
...
+            fsdp_config = {}
+            model = nn.Sequential(
+                nn.Linear(hidden_dim, hidden_dim, device="cuda"),
+                nn.ReLU(),
+                nn.Linear(hidden_dim, hidden_dim, device="cuda"),
+                nn.ReLU(),
+                nn.Linear(hidden_dim, hidden_dim, device="cuda"),
+            )
+            fully_shard(model, reshard_after_forward=True, **fsdp_config)
+            optim = torch.optim.SGD(model.parameters(), lr=1e-6)
```
[the remark this answers is not in the document]
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:197]: changed to `1e-2`

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 6]
```diff
@@ -100,6 +113,172 @@ def f(x):
...
+                    )
+                else:
+                    maybe_compiled_autograd_ctx = contextlib.nullcontext()
+                with maybe_compiled_autograd_ctx:
+                    out = model(inp)
+                    loss = out.sum()
+                    losses.append(loss.item())
+                    loss.backward()
+                optim.step()
+                torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:149]: removed

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py]: Loss seems to go into `NaN` if learning rate is `1e-2`. I changed LR to `1e-4` and also changed `torch.allclose` tolerance to `rtol=1e-5, atol=1e-8`, to ensure that the optimizer steps are visible.

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 7]
```diff
@@ -100,6 +113,154 @@ def f(x):
...
+
+        def model_init_fn():
+            torch.manual_seed(self.rank)
+            fsdp_config = {}
+            mesh = init_device_mesh("cuda", (self.world_size,))
+            model_args = ModelArgs(vocab_size=vocab_size)
+            model = Transformer(model_args)
+            for layer_id, mod in enumerate(model.layers):
+                fully_shard(mod, mesh=mesh, reshard_after_forward=True, **fsdp_config)
+                model.layers[layer_id] = mod
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_compile.py:234]: I think this line is actually not needed :)
```suggestion
```

**Review by awgu — APPROVED** [review]
LGTM!

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py [hunk 8]
```diff
@@ -100,6 +113,172 @@ def f(x):
...
+                    loss = out.sum()
+                    losses.append(loss.item())
+                    loss.backward()
+                optim.step()
+                torch.cuda.synchronize()
+            return losses
+
+        def test_compiled():
+            model, optim = model_init_fn()
+            # FSDP2 does lazy init using 1st run, so run it once to init using eager mode
```
[the remark this answers is not in the document]
Comment (yf225) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:154]: yes - I am thinking of Dynamo proactively doing one eager mode run when it traces FSDP2 (similar to how we handle optimizer init and also LazyModule init in Dynamo). Wonder do you think it would work?

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py]: > changed torch.allclose tolerance to rtol=1e-5, atol=1e-8

oh I thought these were the defaults

> Loss seems to go into NaN if learning rate is 1e-2

this seems kind of surprising 😮  but sounds good

[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py]: maybe can use Adam instead of SGD -- I have not seen any issues with Adam and `lr=1e-2` for MLPs

File: test/distributed/_composable/fsdp/test_fully_shard_compile.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_compile.py:154]: yep that seems good to me
